### PR TITLE
[History] View past attempts

### PR DESCRIPTION
**Pull request by jeffrey-jian**
## PR Description

This PR adds the ability for users to view their past code executions within history.

On code execution, the client saves a snapshot of the current state of the collaboration, which includes:
1. collab id
2. id of user who executed the code
3. code language
4. raw code text
5. code output / test cases and results

This information persists in the `snapshots` table in [Supabase](https://supabase.com/dashboard/project/ebsbkkiwgplgouxksyig/editor/46286?schema=public).

Users can then access their "attempts" within the history view.

Screenshot 2024-11-13 at 10 32 59 PM

Screenshot 2024-11-13 at 10 33 16 PM

### CODE REVIEW

File: project/apps/collaboration-service/src/domain/ports/collaboration.service.ts [hunk 1]
```diff
@@ -95,53 +96,24 @@ export class CollaborationService {
...
-   * Fetches the information of an active collaboration including the selected question data by its unique identifier.
+   * Fetches the information of an collaboration including the selected question data by its unique identifier.
    * @param collabId - The unique identifier of the collaboration to be fetched.
    * @returns A promise that resolves to the collaboration information data transfer object.
-   *          If the collaboration is not active, the promise will resolve to null.
    * @throws Will handle and log any errors that occur during the retrieval process.
    */
 
-  async getActiveCollabInfo(collabId: string): Promise<CollabInfoDto | null> {
+  async getCollabInfo(collabId: string): Promise<CollabInfoDto> {
```
Comment (jeffrey-jian) [inline project/apps/collaboration-service/src/domain/ports/collaboration.service.ts:105]: updated this method to return the collab, even if it's not active. 

used by history to retrieve question and user info

File: project/apps/collaboration-service/src/adapters/controllers/collaboration.controller.ts [hunk 2]
```diff
@@ -29,17 +33,22 @@ export class CollaborationController {
   }
 
   @MessagePattern({ cmd: 'get_collab_info' })
-  async getActiveCollabInfo(@Payload() collabId: string) {
-    return await this.collaborationService.getActiveCollabInfo(collabId);
-  }
-
-  @MessagePattern({ cmd: 'get_collab_history' })
-  async getCollabDocument(@Payload() collabId: string) {
```
Comment (jeffrey-jian) [inline project/apps/collaboration-service/src/adapters/controllers/collaboration.controller.ts:37]: removed getCollabDocument method, collab document will be retrieved in the new `getAttempts` method.

File: project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts [hunk 3]
```diff
@@ -312,84 +317,73 @@ export class CollaborationSupabase implements CollaborationRepository {
...
   }
 
-  async fetchCollabInfo(collabId: string): Promise<CollabInfoDto> {
+  async fetchCollabInfo(collabId: string): Promise<CollabInfoDto | null> {
     const collab = await this.findById(collabId);
     if (!collab) {
-      throw new NotFoundException(
-        `Collaboration with id ${collabId} not found`,
-      );
+      return null;
```
Comment (jeffrey-jian) [inline project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts:323]: not throwing errors if collab cannot be found.

File: project/apps/collaboration-service/src/domain/ports/collaboration.repository.ts [hunk 4]
```diff
@@ -83,20 +87,20 @@ export abstract class CollaborationRepository {
...
 
   /**
-   * Fetches the collaboration information and associated document for a given collaboration id.
-   * @param id The unique identifier of the collaboration to fetch information for.
-   * @returns A promise that resolves to the collaboration information with document data transfer object.
+   * Updates the language of a collaboration by its unique identifier.
+   * @param collabLanguageData The data transfer object containing the details of the language update.
+   * @returns A promise that resolves to the updated collaboration data transfer object.
    */
-  abstract fetchCollabInfoWithDocument(
```
Comment (jeffrey-jian) [inline project/apps/collaboration-service/src/domain/ports/collaboration.repository.ts:97]: removed unused method

File: project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts [hunk 5]
```diff
@@ -312,84 +317,73 @@ export class CollaborationSupabase implements CollaborationRepository {
...
     }
 
     const { question_id, user1_id, user2_id } = collab;
-    try {
```
Comment (jeffrey-jian) [inline project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts:324]: removed try-catch block here, was originally resulting in throwing within try and catching thrown error in catch

File: project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts [hunk 6]
```diff
@@ -312,84 +317,73 @@ export class CollaborationSupabase implements CollaborationRepository {
...
+    };
 
-      return collabInfoData;
-    } catch (error) {
-      throw new Error(`Failed to fetch collaboration info: ${error}`);
-    }
+    return collabInfoData;
   }
 
-  async fetchCollabInfoWithDocument(
```
Comment (jeffrey-jian) [inline project/apps/collaboration-service/src/adapters/db/collaboration.supabase.ts:375]: removed unused method.

**Review by javinchua — APPROVED** [review]
lgtm

File: project/apps/web/components/history-view/HistoryEditor.tsx [hunk 7]
```diff
@@ -1,65 +1,115 @@
...
+              testResults={testResults!}
+              testCases={testCases}
+            />
+          ) : (
+            <div className="h-full px-6 py-4 overflow-auto whitespace-pre-wrap">
+              {output}
+            </div>
+          )}
+        </div>
+      ) : null}
```
[the remark this answers is not in the document]
Comment (jeffrey-jian) [reply project/apps/web/components/history-view/HistoryEditor.tsx:110]: resolved.

**Pull request merged**